man/ipetoipe.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics, 
.\" respectively.
.TH IPETOIPE 1 "June 27, 2009"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
ipetoipe \- Convert between Ipe file formats
.SH SYNOPSIS
.B ipetoipe
( -pdf | -xml ) { options } \fIinput-file\fP [ \fIoutput-file\fP ]

.SH DESCRIPTION
.PP
\fBipetoipe\fP converts between the two Ipe file formats XML and PDF.
Options are:
.TP
\fB-pdf\fP
convert to PDF format
.TP
\fB-xml\fP
convert to XML format
.TP
\fB-export\fP
do not include Ipe markup in the output file.
.br
\fIWarning:\fP this means that Ipe will not be able to read the
resulting output file. 
.TP
\fB-pages\fP \fIfrom\fP-\fIto\fP
restrict exporting to PDF to this page range.  This implies  the
\fB-export\fP option. 
.TP
\fB-view\fP \fIpage\fP-\fIview\fP
only export this single view from the document.  This implies  the
\fB-export\fP option. 
.TP
\fB-markedview\fP
create a PDF page only for the marked views of marked Ipe pages. This
is for instance convenient for making handouts for slides used in 
a presentation. This implies the
\fB-export\fP option. 
.TP
\fB-runlatex\fP
run Latex even for XML output. This has the effect of including the
dimensions of each text object in the XML file.
.TP
\fB-nozip\fP
do not compress streams in PDF or Postscript output.

.SH ENVIRONMENT VARIABLES

\fBipetoipe\fP respects the following environment variables:

.TP
\fBIPELATEXDIR\fP
the directory where \fBipetoipe\fP runs Pdflatex.

.TP
\fBIPELATEXPATH\fP
the directory containing pdflatex and xelatex.  If not set, they are
searched for on the path.

.TP
\fBIPEDEBUG\fP
set this to 1 for debugging output.

.SH AUTHOR
Otfried Cheong

.SH REPORTING BUGS
.ad l
Please report bugs using Ipe bugzilla at
.I "http://ipe.otfried.org/bugzilla.html"

.SH SEE ALSO
.ad l
Ipe is documented fully in
.IR "The Ipe Manual" ,
which is available online at
.I "http://ipe.otfried.org/manual/manual.html" .


doc/news.txt
----------------------------------------------------------------------

Ipe 7.2.28

 * Add "sequential text" option.

 * Fix centering of title in presentation stylesheet (bug #464).

----------------------------------------------------------------------

Ipe 7.2.27 (2023-05-05)

 * On Linux, migrate to Qt6 (#457).  Thanks Henri!

 * Change behaviour of transform tools to never switch the current
   selection (#459).

 * Action to deselect everything on current page.

 * Align center/H center/V center now aligns to the page if
   only one object selected (#449).

 * Fix absolute autosave paths (bug #433).

 * Add /opt/homebrew/bin to path on MacOS (#440).

 * Fix missing 'stroke-opacity' and 'opacity' attributes in
   'ipe.dtd' (#67).

 * Fix crash when 'recent_files_lua' ends up empty (bug #455).

 * Fix missing page numbers when there is no text (bug #450).

 * Don't lose group's link action during group edit (bug #456).

----------------------------------------------------------------------

Ipe 7.2.26 (2022-07-16)

 * Allow to store autosave file in same directory as document
   itself.  This is now also the default. (#426)

 * Circular graph edges no longer get broken when moving graph
   nodes (bug #424).

 * Make it possible to remap left_command on MacOS (bug #431).

 * Use PKG_CONFIG environment variable (#429).

----------------------------------------------------------------------

Ipe 7.2.25 (2022-06-19) Denmark Edition

 * Add spell-checking to text entry on Linux.

 * Change presentation style sheet to have 16:9 aspect ratio (#385).

 * Many more configuration options in prefs.canvas_style (#400 &
   #406).

 * Remove box around links in PDF output (#397).

 * Add 'script' text size (#410).

 * Use most-recently used directory in "Save" and "Save As" (#394).

 * Maintain a single custom string on each Ipe object. It's not
   used by Ipe itself, but can be used by external software or
   ipelets. (#409)

 * Add page transition effects from PDF 1.5 (Push, Cover, Uncover,
   Fade).

 * Adjust highlighting text colors in dark theme.

 * Fix crash when compiling with -D_GLIBCXX_ASSERTIONS (#387).

 * Try to simply use UTF-8 in 'runlatex.bat'.  This should work
   at least on Windows-10 (bug #338, second case).

 * Upgrade Freetype in Windows package (fixes bug #411).

 * Show error message when a style sheet mentioned in prefs.styles
   cannot be read (bug #405).

 * Reuse PDF text resources for identical text objects (#412).

 * Fix losing primary selection when deselecting using rectangle drag
   (bug #423).

----------------------------------------------------------------------

Ipe 7.2.24 (2021-04-07)

 * Fix a crash on MacOS 11 El Sur by applying a patch to cairo
   (bug #370).

 * Switch from CAIRO_FILTER_FAST to CAIRO_FILTER_GOOD for decent
   display (essential when filming from the canvas).

 * Make "Select by type/attribute" ipelet work when selecting on an
   absolute color (#380).

 * Fix semi-transparent bitmaps as well as bitmaps with color keys not
   working in Ipepresenter (bug #9).

 * Fix exporting to EPS when part of the drawing lies outside the
   paper (bug #369).

 * Remember online latex service URL in a different file to avoid
   problems when using old and new Ipe versions.

 * Improve calling of external editor on Qt, so that keyboard focus
   is given to the external editor.

 * Raise Ipe window after returning from external editor on Qt, in
   case the editor program is still running.

 * New preference 'freeze_in_latex' to let the Ipe UI freeze during a
   Latex run.

 * New shell script 'iperender-par' that allows to run several
   instances of iperender in parallel (e.g. as part of a parallel
   make) (#375).

 * Add Lua API to access (Latex) dimensions of text object.

 * Silently ignore non-integer grid sizes (#372).

 * On MacOS 11, return to the classic toolbar underneath the title
   bar, because the new unified look leaves most tool buttons off the
   screen (bug #376).

----------------------------------------------------------------------

Ipe 7.2.23 (2020-12-24)

 * Shortcut for pretty display.

 * A maximum on the number of recent files remembered.

 * Improve scaling of toolbar buttons on El Sur, and add labels for
   the toolbar overflow menu.
 
 * Remove sources of libspiro from Ipe releases.

 * Remove doxygen documentation from Ipe release, run doxygen as part
   of the build process.

----------------------------------------------------------------------

Ipe 7.2.22 (2020-12-20)

 * Recent file menu (#322).

 * Update PDF resources on main thread to avoid race condition (bugs
   #354, #359).

 * Can exclude pages from PDF output by adding a layer "NOPDF" to the
   page. The layer does not need to be visible. (#362).  

 * Use a better bitmap checksum that is less likely to cause bitmaps
   to be reordered.  Sort mappings of symbolic attributes to make them
   deterministic (#356).

 * Display Type3 fonts with a unique representation to make it easy to
   diagnose the problem.

----------------------------------------------------------------------

Ipe 7.2.21 (2020-10-26)

 * Fix hanging with message "Running Latex" on Windows (bug #338).

 * Ability to hide the properties panel/layer list from the UI on
   macOS, for use on Zoom etc (#350).
 
 * Remove NoFocus policy from selectors in Qt.  This used to be
   necessary to keep shortcuts working when the selector receives
   focus, but that's no longer necessary (tested in Qt 5.9).
   Fixes selectors not working on Qt 5.15 (bug #341).

 * Fix incorrect update of grid size and angle size selector when
   style sheet is changed (bug #344).

 * Fix crash in 'iperender -pdf' when the document uses
   \includegraphics on a jpg file (bug #342).

 * Fix snapping to midpoint of closing segment of closed paths (bug
   #346).

 * Fix Fifi remaining invisible after switching back from ink mode
   (bug #347).

 * Escape key now closes page/view selectors and sorters on Windows
   (bug #349).


src/ipetoipe/ipetoipe.cpp
// --------------------------------------------------------------------
// ipetoipe
// --------------------------------------------------------------------
/*

    This file is part of the extensible drawing editor Ipe.
    Copyright (c) 1993-2023 Otfried Cheong

    Ipe is free software; you can redistribute it and/or modify it
    under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 3 of the License, or
    (at your option) any later version.

    As a special exception, you have permission to link Ipe with the
    CGAL library and distribute executables, as long as you follow the
    requirements of the Gnu General Public License in regard to all of
    the software in the executable aside from CGAL.

    Ipe is distributed in the hope that it will be useful, but WITHOUT
    ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
    or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
    License for more details.

    You should have received a copy of the GNU General Public License
    along with Ipe; if not, you can find it at
    "http://www.gnu.org/copyleft/gpl.html", or write to the Free
    Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

*/

#include "ipedoc.h"
#include <cstdlib>
#include <cstring>

using ipe::Document;
using ipe::String;
using ipe::FileFormat;
using ipe::SaveFlag;

static int topdf(Document *doc, String src, String dst, uint32_t flags,
		 int fromPage = -1, int toPage = -1, int viewNo = -1)
{
  int res = doc->runLatex(src);
  if (res) return res;

  bool result = false;
  if (viewNo >= 0) {
    result = doc->exportView(dst.z(), FileFormat::Pdf, flags, fromPage, viewNo);
  } else if (toPage >= 0) {
    result = doc->exportPages(dst.z(), flags, fromPage, toPage);
  } else {
    result = doc->save(dst.z(), FileFormat::Pdf, flags);
  }
  if (!result) {
    fprintf(stderr, "Failed to save or export document!\n");
    return 1;
  }

  if (flags & SaveFlag::Export)
    fprintf(stderr, "Warning: the exported file contains no Ipe markup.\n"
	    "It cannot be read by Ipe - make sure you keep the original!\n");
  return 0;
}

static void usage()
{
  fprintf(stderr,
	  "Usage: ipetoipe ( -xml | -pdf ) <options> "
	  "infile [ outfile ]\n"
	  "Ipetoipe converts between the different Ipe file formats.\n"
	  " -export      : output contains no Ipe markup.\n"
	  " -pages <n-m> : export only these pages (implies -export).\n"
	  " -view <p-v>  : export only this view (implies -export).\n"
	  " -markedview  : export only marked views on marked pages (implies -export).\n"
	  "     If all views of a marked page are unmarked, the last view is exported.\n"
	  "     This is convenient to make handouts for slides.\n"
	  " -runlatex    : run Latex even for XML output.\n"
	  " -nozip       : do not compress PDF streams.\n"
	  " -keepnotes   : save page notes as PDF annotations even when exporting.\n"
	  "Pages can be specified by page number or by section title.\n"
	  );
  exit(1);
}

int main(int argc, char *argv[])
{
  ipe::Platform::initLib(ipe::IPELIB_VERSION);

  // ensure at least two arguments (handles -help as well :-)
  if (argc < 3)
    usage();

  FileFormat frm = FileFormat::Unknown;
  if (!strcmp(argv[1], "-xml"))
    frm = FileFormat::Xml;
  else if (!strcmp(argv[1], "-pdf"))
    frm = FileFormat::Pdf;

  if (frm == FileFormat::Unknown)
    usage();

  uint32_t flags = SaveFlag::SaveNormal;
  bool runLatex = false;
  const char *pages = nullptr;
  const char *view = nullptr;
  int i = 2;

  String infile;
  String outfile;

  while (i < argc) {

    if (!strcmp(argv[i], "-export")) {
      flags |= SaveFlag::Export;
      ++i;
    } else if (!strcmp(argv[i], "-view")) {
      flags |= SaveFlag::Export;
      if (i + 1 == argc)
	usage();
      view = argv[i+1];
      i += 2;
    } else if (!strcmp(argv[i], "-pages")) {
      flags |= SaveFlag::Export;
      if (i + 1 == argc)
	usage();
      pages = argv[i+1];
      i += 2;
    } else if (!strcmp(argv[i], "-markedview")) {
      flags |= SaveFlag::MarkedView;
      flags |= SaveFlag::Export;
      ++i;
    } else if (!strcmp(argv[i], "-runlatex")) {
      runLatex = true;
      ++i;
    } else if (!strcmp(argv[i], "-nozip")) {
      flags |= SaveFlag::NoZip;
      ++i;
    } else if (!strcmp(argv[i], "-keepnotes")) {
      flags |= SaveFlag::KeepNotes;
      ++i;
    } else {
      // last one or two arguments must be filenames
      infile = argv[i];
      ++i;
      if (i < argc) {
	outfile = argv[i];
	++i;
      }
      if (i != argc)
	usage();
    }
  }

  if (infile.empty())
    usage();

  if ((flags & SaveFlag::Export) && frm == FileFormat::Xml) {
    fprintf(stderr, "-export only available with -pdf.\n");
    exit(1);
  }

  if (pages && frm != FileFormat::Pdf) {
    fprintf(stderr, "-pages only available with -pdf.\n");
    exit(1);
  }

  if (pages && view) {
    fprintf(stderr, "cannot specify both -pages and -view.\n");
    exit(1);
  }

  if (outfile.empty()) {
    outfile = infile;
    String ext = infile.right(4);
    if (ext == ".ipe" || ext == ".pdf" || ext == ".xml")
      outfile = infile.left(infile.size() - 4);
    switch (frm) {
    case FileFormat::Xml:
      outfile += ".ipe";
      break;
    case FileFormat::Pdf:
      outfile += ".pdf";
    default:
      break;
    }
    if (outfile == infile) {
      fprintf(stderr, "Cannot guess output filename.\n");
      exit(1);
    }
  }

  std::unique_ptr<Document> doc(Document::loadWithErrorReport(infile.z()));

  if (!doc)
    return 1;

  fprintf(stderr, "Document %s has %d pages (%d views)\n",
	  infile.z(), doc->countPages(), doc->countTotalViews());

  // parse pages and view
  int fromPage = -1;
  int toPage = -1;
  int viewNo = -1;

  if (pages) {
    String p(pages);
    int j = p.find('-');
    if (j >= 0) {
      fromPage = (j > 0) ? doc->findPage(p.left(j)) : 0;
      toPage = (j < p.size() - 1) ? doc->findPage(p.substr(j+1)) : doc->countPages() - 1;
    }
    if (fromPage < 0 || fromPage > toPage) {
      fprintf(stderr, "incorrect -pages specification.\n");
      exit(1);
    }
  } else if (view) {
    String v(view);
    int j = v.find('-');
    if (j > 0) {
      fromPage = doc->findPage(v.left(j));
      if (fromPage >= 0)
	viewNo = doc->page(fromPage)->findView(v.substr(j+1));
    }
    if (fromPage < 0 || viewNo < 0) {
      fprintf(stderr, "incorrect -view specification.\n");
      exit(1);
    }
  }

  char buf[64];
  sprintf(buf, "ipetoipe %d.%d.%d",
	  ipe::IPELIB_VERSION / 10000,
	  (ipe::IPELIB_VERSION / 100) % 100,
	  ipe::IPELIB_VERSION % 100);
  Document::SProperties props = doc->properties();
  props.iCreator = buf;
  doc->setProperties(props);

  switch (frm) {
  case FileFormat::Xml:
    if (runLatex)
      return topdf(doc.get(), infile, outfile, flags);
    else
      doc->save(outfile.z(), FileFormat::Xml, SaveFlag::SaveNormal);
  default:
    return 0;

  case FileFormat::Pdf:
    return topdf(doc.get(), infile, outfile, flags, fromPage, toPage, viewNo);
  }
}

// --------------------------------------------------------------------
